### DISCUSSION TITLE
Issues with Using SSD1306 and MPU6050 Simultaneously on ESP32

### DISCUSSION THREAD

**Discussion by angbas06** [community/community#136555] [Programming Help]
### Body

Hi everyone,

I’ve been working on programming an ESP32 with an integrated SSD1306 OLED display and an MPU6050 (accelerometer and gyroscope). I’ve managed to get both working individually using PlatformIO—displaying a "Hello World" message on the screen and reading gyroscope data on the serial terminal. However, I’m having trouble using both simultaneously.

It seems like I can only use either the SSD1306 or the MPU6050 at a time, but not both together to show data on the OLED display. Does anyone have any advice or suggestions on how to achieve this?

Here’s the code I’m currently using:

#include <Adafruit_MPU6050.h>
#include <Adafruit_Sensor.h>
#include <Wire.h>

Adafruit_MPU6050 mpu;

void setup(void) {
  Serial.begin(115200);
  while (!Serial)
    delay(10); // will pause Zero, Leonardo, etc until serial console opens

  Serial.println("Adafruit MPU6050 test!");

  // Try to initialize!
  if (!mpu.begin()) {
    Serial.println("Failed to find MPU6050 chip");
    while (1) {
      delay(10);
    }
  }
  Serial.println("MPU6050 Found!");

  mpu.setAccelerometerRange(MPU6050_RANGE_8_G);
  Serial.print("Accelerometer range set to: ");
  switch (mpu.getAccelerometerRange()) {
  case MPU6050_RANGE_2_G:
    Serial.println("+-2G");
    break;
  case MPU6050_RANGE_4_G:
    Serial.println("+-4G");
    break;
  case MPU6050_RANGE_8_G:
    Serial.println("+-8G");
    break;
  case MPU6050_RANGE_16_G:
    Serial.println("+-16G");
    break;
  }
  mpu.setGyroRange(MPU6050_RANGE_500_DEG);
  Serial.print("Gyro range set to: ");
  switch (mpu.getGyroRange()) {
  case MPU6050_RANGE_250_DEG:
    Serial.println("+- 250 deg/s");
    break;
  case MPU6050_RANGE_500_DEG:
    Serial.println("+- 500 deg/s");
    break;
  case MPU6050_RANGE_1000_DEG:
    Serial.println("+- 1000 deg/s");
    break;
  case MPU6050_RANGE_2000_DEG:
    Serial.println("+- 2000 deg/s");
    break;
  }

  mpu.setFilterBandwidth(MPU6050_BAND_21_HZ);
  Serial.print("Filter bandwidth set to: ");
  switch (mpu.getFilterBandwidth()) {
  case MPU6050_BAND_260_HZ:
    Serial.println("260 Hz");
    break;
  case MPU6050_BAND_184_HZ:
    Serial.println("184 Hz");
    break;
  case MPU6050_BAND_94_HZ:
    Serial.println("94 Hz");
    break;
  case MPU6050_BAND_44_HZ:
    Serial.println("44 Hz");
    break;
  case MPU6050_BAND_21_HZ:
    Serial.println("21 Hz");
    break;
  case MPU6050_BAND_10_HZ:
    Serial.println("10 Hz");
    break;
  case MPU6050_BAND_5_HZ:
    Serial.println("5 Hz");
    break;
  }

  Serial.println("");
  delay(100);
}

void loop() {

  /* Get new sensor events with the readings */
  sensors_event_t a, g, temp;
  mpu.getEvent(&a, &g, &temp);

  /* Print out the values */
  Serial.print("Acceleration X: ");
  Serial.print(a.acceleration.x);
  Serial.print(", Y: ");
  Serial.print(a.acceleration.y);
  Serial.print(", Z: ");
  Serial.print(a.acceleration.z);
  Serial.println(" m/s^2");

  Serial.print("Rotation X: ");
  Serial.print(g.gyro.x);
  Serial.print(", Y: ");
  Serial.print(g.gyro.y);
  Serial.print(", Z: ");
  Serial.print(g.gyro.z);
  Serial.println(" rad/s");

  Serial.print("Temperature: ");
  Serial.print(temp.temperature);
  Serial.println(" degC");

  Serial.println("");
  delay(500);
}

Thanks in advance for any help!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by angbas06** [comment]
I apologize, I’ve lost the code for both devices. However, it wasn’t working as expected. The code above is only for the gyroscope.

**Comment by danialtavakoli** [comment]
To use the SSD1306 OLED display and MPU6050 sensor together on the ESP32, ensure they have different I2C addresses and are correctly initialized in your code. Integrate both in the `setup` function, and in the `loop`, fetch data from the MPU6050 and display it on the OLED. If issues persist, check your power supply and consider adding pull-up resistors to the I2C lines. This should allow both devices to work simultaneously without conflicts.

**Reply by angbas06** [reply]
I have another issue now. Initially, I created a program to display random numbers on the screen. The display has the address 0x3C and is connected to pins 4 and 5. Later, I tested the gyroscope on the serial monitor. The gyroscope has the address 0x68 and is connected to pins 21 and 22. However, when I try to use both addresses separately in the code, it doesn't work. Combining them doesn’t work either. When I perform an I2C scan, I only see the address 0x68. The display, when addressed as 0x3, works fine. I have no idea what to do. Thank you for your response.

**Reply by danialtavakoli** [reply]
It seems like the I2C bus isn't detecting both devices properly. The fact that only address 0x68 shows up in your scan suggests a possible conflict or wiring issue. Ensure that both devices are correctly connected to the correct I2C pins (typically, the same SDA and SCL pins for both). Since you’re seeing 0x68 but not 0x3C, double-check the wiring for the OLED display. Also, confirm that both devices are correctly powered. You might want to try connecting both devices to the same I2C bus on the same pins (e.g., using pins 21 and 22 for both). If the problem persists, consider using a logic analyzer to troubleshoot the I2C communication further. If this response was helpful, feel free to mark it as the answer for support.

**Reply by angbas06** [reply]
Unfortunately, the board already has the screen soldered on, and I can't see exactly how and where it is connected. Specifically, this is a TTGO board, and the screen is soldered to a point labeled SCL and SDA (I assume it’s on a tin base). However, using this line of code: SSD1306Wire display(0x3c, 5, 4); activates the display, so I assume it is connected to those two pins. On the other hand, I soldered the MPU6050 myself, and it is powered correctly, as each component works individually, but not when used together.

**Reply by angbas06** [reply]
![ttgo_p0](https://github.com/user-attachments/assets/c98d680d-2fe3-4533-8065-ab34a91c3932)
![ttgo_pinout](https://github.com/user-attachments/assets/a0159ea9-2a1a-4884-b495-154d65deff0a)

**Comment by danialtavakoli** [comment]
Given that you're working with a TTGO board where the screen is already soldered and connected to specific pins (SCL and SDA), and the SSD1306 display is activated with `SSD1306Wire display(0x3c, 5, 4);`, it's likely that these pins are fixed for the OLED display. The MPU6050 you added may be conflicting on the I2C bus or the pin assignments!

Since the I2C addresses seem to be correct (0x3C for the OLED and 0x68 for the MPU6050), the issue likely lies in how the devices are sharing the I2C bus. Both devices should be on the same I2C bus (same SDA and SCL lines). On your TTGO board, these lines might already be hardwired to certain pins.

steps for you:

1. **Confirm the I2C Pins:** Ensure that both the OLED and MPU6050 are indeed connected to the same I2C pins. If the display is on pins 5 and 4 (likely SDA and SCL respectively), the MPU6050 should also be connected to these same pins.

2. **Check for Pull-up Resistors:** The I2C bus requires pull-up resistors on both the SDA and SCL lines. If both devices have built-in pull-ups, they might interfere with each other. Alternatively, if neither has pull-ups, the bus might not function correctly.

3. **Ensure Proper Power Supply:** Since the MPU6050 works individually, ensure that there’s enough power supplied when both the display and sensor are connected. Sometimes power issues can cause devices not to be detected correctly on the I2C bus.

4. **Test Different Pins if Possible:** If the TTGO board allows, try connecting the MPU6050 to different GPIO pins that can act as alternate I2C bus lines (if configurable in code).

If after checking these points the problem persists, it might be helpful to use a logic analyzer to see the communication on the I2C bus, or to refer to the TTGO board’s specific documentation to understand how the I2C lines are wired internally.

**Reply by angbas06** [reply]
Unfortunately, I’m not at work right now and don’t have the physical means to do it. However, tomorrow I will try to desolder the MPU6050 and resolder it to pins 4 and 5 so that it will be forced to share that channel with the OLED display.

**Reply by danialtavakoli** [reply]
Good luck!

**Reply by angbas06** [reply]
Hi, I just tried using different pins, but the sensor still isn’t working. So, I re-soldered the sensor to pins 21 and 22 because it worked on the serial monitor with those pins. You mentioned adding pull-up resistors, right? But how can I check if the OLED screen already has pull-up resistors built in?
